File: src/org/nwapw/abacus/tree/NumberReducer.java
package org.nwapw.abacus.tree;

import org.nwapw.abacus.function.Function;
import org.nwapw.abacus.number.NumberInterface;
import org.nwapw.abacus.plugin.PluginManager;

/**
 * A reducer implementation that turns a tree into a single number.
 * This is not always guaranteed to work.
 */
public class NumberReducer implements Reducer<NumberInterface> {

    /**
     * The plugin manager from which to draw the functions.
     */
    private PluginManager manager;

    /**
     * Creates a new number reducer with the given plugin manager.
     * @param manager the plugin manager.
     */
    public NumberReducer(PluginManager manager){
        this.manager = manager;
    }

    @Override
    public NumberInterface reduceNode(TreeNode node, Object... children) {
        if(node instanceof NumberNode) {
            return ((NumberNode) node).getNumber();
        } else if(node instanceof OpNode){
            NumberInterface left = (NumberInterface) children[0];
            NumberInterface right = (NumberInterface) children[1];
            Function function = manager.operatorFor(((OpNode) node).getOperation()).getFunction();
            if(function == null) return null;
            return function.apply(left, right);
        } else if(node instanceof FunctionNode){
            NumberInterface[] convertedChildren = new NumberInterface[children.length];
            for(int i = 0; i < convertedChildren.length; i++){
                convertedChildren[i] = (NumberInterface) children[i];
            }
            Function function = manager.functionFor(((FunctionNode) node).getFunction());
            if(function == null) return null;
            return function.apply(convertedChildren);
        }
        return null;
    }

}

File: core/src/main/kotlin/org/nwapw/abacus/tree/standard/NumberReducer.kt
package org.nwapw.abacus.tree.standard

import org.nwapw.abacus.Abacus
import org.nwapw.abacus.context.EvaluationContext
import org.nwapw.abacus.exception.NumberReducerException
import org.nwapw.abacus.exception.ReductionException
import org.nwapw.abacus.number.NumberInterface
import org.nwapw.abacus.tree.Reducer
import org.nwapw.abacus.tree.nodes.*

class NumberReducer(val abacus: Abacus, context: EvaluationContext) : Reducer<NumberInterface> {

    val context = context.mutableSubInstance()

    init {
        this.context.reducer = this
    }

    override fun reduceNode(treeNode: TreeNode, vararg children: Any): NumberInterface {
        val promotionManager = abacus.promotionManager
        return when(treeNode){
            is NumberNode -> {
                context.inheritedNumberImplementation.instanceForString(treeNode.number)
            }
            is VariableNode -> {
                val variable = context.getVariable(treeNode.variable)
                if(variable != null) return variable
                val definition = context.getDefinition(treeNode.variable)
                if(definition != null) return definition.reduce(this)
                throw NumberReducerException("variable is not defined.")
            }
            is NumberUnaryNode -> {
                val child = children[0] as NumberInterface
                context.numberImplementation = abacus.pluginManager.interfaceImplementationFor(child.javaClass)
                abacus.pluginManager.operatorFor(treeNode.operation)
                        .apply(context, child)
            }
            is NumberBinaryNode -> {
                val left = children[0] as NumberInterface
                val right = children[1] as NumberInterface
                val promotionResult = promotionManager.promote(left, right)
                context.numberImplementation = promotionResult.promotedTo
                abacus.pluginManager.operatorFor(treeNode.operation).apply(context, *promotionResult.items)
            }
            is NumberFunctionNode -> {
                val promotionResult = promotionManager
                        .promote(*children.map { it as NumberInterface }.toTypedArray())
                context.numberImplementation = promotionResult.promotedTo
                abacus.pluginManager.functionFor(treeNode.callTo).apply(context, *promotionResult.items)
            }
            is TreeValueUnaryNode -> {
                abacus.pluginManager.treeValueOperatorFor(treeNode.operation)
                        .apply(context, treeNode.applyTo)
            }
            is TreeValueBinaryNode -> {
                abacus.pluginManager.treeValueOperatorFor(treeNode.operation)
                        .apply(context, treeNode.left, treeNode.right)
            }
            is TreeValueFunctionNode -> {
                abacus.pluginManager.treeValueFunctionFor(treeNode.callTo)
                        .apply(context, *treeNode.children.toTypedArray())
            }
            else -> throw ReductionException("unrecognized tree node.")
        }
    }

}